cpp/src/gandiva/function_ir_builder.cc

```diff
@@ -0,0 +1,59 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "gandiva/function_ir_builder.h"
+
+namespace gandiva {
+
+llvm::Value* FunctionIRBuilder::BuildIfElse(llvm::Value* condition,
+                                            llvm::Type* return_type,
+                                            std::function<llvm::Value*()> then_func,
```

Comment:
I'm curious: what does it bring to pass a function here, rather than directly the corresponding `llvm::Value`?

Reply:
you are right - I'll fix this.

Reply:
sorry, I agreed too soon - these have to be functions. The code-generation for the if/else blocks must happen after the corresponding BasicBlock has started i.e after the SetInsertPoint call. Otherwise, there is no conditional evaluation i.e both if/else blocks will get evaluated.

Tried the following simple case:
   _x_value = BuildIfElse(lt_max_precision, Add11(x_value), Sub11(x_value));_

**without the functions (i.e using direct lvalue, I get)**
```
if:
  %3 = add i128 %member, 11
  %4 = sub i128 %member, 11
  br i1 %2, label %then, label %else

exit:                                             ; preds = %merge
  ret void

then:                                             ; preds = %entry
  br label %merge

else:                                             ; preds = %entry
  br label %merge

merge:                                            ; preds = %else, %then
  %res_value = phi i128 [ %3, %then ], [ %4, %else ]
```
**with the lambda functions**

 ```
 %2 = icmp slt i32 %member1, 38
  br i1 %2, label %then, label %else

exit:                                             ; preds = %merge
  ret void

then:                                             ; preds = %entry
  %3 = add i128 %member, 11
  br label %merge

else:                                             ; preds = %entry
  %4 = sub i128 %member, 11
  br label %merge

merge:                                            ; preds = %else, %then
  %res_value = phi i128 [ %3, %then ], [ %4, %else ]
```